text/0000-overload-logic-operators.md

```diff
@@ -0,0 +1,211 @@
+    Ok(2) || 1;               // == 2
+
+    Ok(2) || Ok("hello");  // Error: LogicalOr<Result<&str, _>> not implemented for Result<i32, _>
+    Ok(2) || 2 || 3;       // Error: LogicalOr<i32> is not implemented for i32
+}
+```
+
+The feature should be thought about as moving the logic from methods and into the current function.
+It maps very seamlessly from using the methods and is equivalent in use without having to worry about
+the naming convention of the short circuit methods. The same mental state of short circuit from
+bools applies directly, without having any recourse to "truthiness" which is not a desirable trait.
+
+This RFC also proposes to deprecate the `.or(...)`, `.or_with(|| ...)`, `.and(...)`, `.and_with(||
+...)`, `.unwrap_or(...)`, and `.unwrap_or_with(|| ...)` methods on `Option<T>` and `.or(...)`,
+`.or_with(|| ...)`, `.unwrap_or(...)`, and `.unwrap_or_with(|| ...)` methods on `Result<T, E>` since
+using this feature renders them unneeded.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The basis of this proposal are the two new traits. These should be implemented along the same lines
+as other operator traits. However, the desugaring should follow the `match` output in the previous
+section so as to obtain the desired short circuit operation.
+
+Once the traits have been implemented, several trait implementations should be added to the std library and the methods marked as deprecated.
+
+```rust
+impl LogicalOr<Option<T>> for Option<T> {
+    type Output = Self;
+    ...
+}
+
+impl trait LogicalOr<T> for Option<T> {
+    type Output = T;
+}
+
+impl trait LogicalAnd<Option<T>> for Option<T> {
+    type Output = Self;
+    ...
+}
+
+impl trait LogicalOr<Result<T, E>> for Result<T, E> {
+    type Output = Self;
+    ...
+}
+
+impl trait LogicalOr<T> for Result<T, E> {
+    type Output = T;
+}
+
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. Leaves the `||` and the `&&` as not strictly boolean operators, which might hurt readability
+2. Could lead to similarities to C++'s Operator bool() which => truthiness and is undesirable.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- This design is the best because it does not rely on other traits, which are how the other operator traits work in Rust. It has the minimum overhead since it does not rely on closures.
+- Two alternatives were discussed and shown to be inferior in either performance, explicitness, or design.
+- The first being "truthiness" conversion trait and then automatically allowing `||` and `&&` if both that trait and the `BitOr` or `BitAnd` traits were also implemented. This was discarded because we did not want to go down the javascript route of auto converting things to boolean (Rust already does not allow non-bools in check expressions) and auto traits are not something that Rust has so that was another reason not to go down this route.
```

Comment:
These two alternatives should be nested under the "Two alternatives were discussed" point